### ISSUE TITLE
Release of object pointer is omitted

### ISSUE THREAD

**Issue by RolfKal** [robertguetzkow/libtrashcan#10]
**Describe the bug**
In the Windows code, after passing the pSI item pointer to the IFileOperation:DeleteItem() function, the pSI object should be released with IShellItem:Release(pSI), This should happen immediately after the IFileOperation:DeleteItem() call and independent if that call was successful or not, since that function should have done an AddRef() on the object pointer until it doesn't need it anymore.

**libtrashcan version**
commit (https://github.com/robertguetzkow/libtrashcan/commit/c8b05a132cceeb0e0de0b192a317ff9cc3c7e895) on Oct 7, 2019

**Operating system**
Windows, any version

**Comment by robertguetzkow** [comment]
Thank you for the report and apologies for the late reply! The reference count for `IShellItem` is indeed not lowered. I will investigate how to solve this properly in C.

**Comment by robertguetzkow** [comment]
In order to decrement the reference count `pSI->lpVtbl->Release(pSI);` is used in C. The decrement does not have to happen immediately after `pfo->lpVtbl->DeleteItem(pfo, pSI, NULL);`. Therefore the suggested fix is:

```diff
@@ -175,10 +175,13 @@ int trashcan_soft_delete_core(const wchar_t *path, bool init_com)
        if (FAILED(hr)) { HANDLE_ERROR(status, LIBTRASHCAN_PARSE, error_2) }

        hr = pfo->lpVtbl->DeleteItem(pfo, pSI, NULL);
-       if (FAILED(hr)) { HANDLE_ERROR(status, LIBTRASHCAN_SETOP, error_2) }
+       if (FAILED(hr)) { HANDLE_ERROR(status, LIBTRASHCAN_SETOP, error_3) }

        hr = pfo->lpVtbl->PerformOperations(pfo);
-       if (FAILED(hr)) { HANDLE_ERROR(status, LIBTRASHCAN_EXECOP, error_2) }
+       if (FAILED(hr)) { HANDLE_ERROR(status, LIBTRASHCAN_EXECOP, error_3) }
+
+error_3:
+       pSI->lpVtbl->Release(pSI);

 error_2:
        pfo->lpVtbl->Release(pfo);
```

References:
https://docs.microsoft.com/en-us/office/client-developer/outlook/mapi/implementing-iunknown-in-c
https://docs.microsoft.com/de-de/windows/win32/learnwin32/managing-the-lifetime-of-an-object

**Comment by robertguetzkow** [comment]
@RolfKal It would be great, if you could double check the suggested fix provided in the [2022-04-12-patch-reference-counting](https://github.com/robertguetzkow/libtrashcan/tree/2022-04-12-patch-reference-counting) branch.

**Comment by robertguetzkow** [comment]
This issue should be fixed by 045fdc15f68d57c4e03408389fa005076fd1f733

**Comment by RolfKal** [comment]
While the reference count needs indeed not to be released right away after passing it to the IFileOperation_DeleteItem() function it doesn't hurt to do it anyways. This would have saved you adding another error goto label.

I tend to program defensive, meaning release resources as early as possible, if it doesn't make the code more complex.

**Comment by robertguetzkow** [comment]
That is very valid criticism. The reason why I've placed it there is because the `pfo->lpVtbl->DeleteItem(pfo, pSI, NULL);` does not actually perform the deletion, this happens when `pfo->lpVtbl->PerformOperations(pfo);` is called. Therefore, I was under the impression that decrementing the reference count before performing the operation might not be correct, since the memory `IShellItem *pSI` points to is still needed and should not be freed yet. Based on the documentation it did not seem safe to immediately call `pSI->lpVtbl->Release(pSI);`. Perhaps the concern is unfounded, depending on how `DeleteItem` and `PerformOperations` are implemented (in particular whether `DeleteItem` calls `AddRef`). Do you have a link to documentation that states the reference count can be decremented before performing the operation?

References:
- https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-deleteitem
- https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-performoperations

**Comment by RolfKal** [comment]
This is standard COM operation. An object method receiving an object pointer that will be used after the method has returned always needs to do an explicit AddRef() on that object or it risks that the object gets invalidated in the meantime. It's a very different approach to modern programming languages automatic garbage collection design, but it has the advantage of being rather efficient if the programmer observes the rules that come with it.

That is the entire (well mostly) reason behind the refcounting mechanism in COM. Of course many developers do not understand refcounting and mess up the AddRef() and Release() calls pretty regularly. But this is a Windows shell component and if anyone, then the Microsoft developers should understand how to do it right.

Also note that the API might actually not even have finished when the PerformOperation() returns. This doesn't apply in this specific case if I interpret the MSDN information correctly, but could be a very valid approach, especially if one of the IFileOperationProgressSink parameters would be used.

**Comment by robertguetzkow** [comment]
Yes, that is obviously the idea behind the reference counting for COM and one could expect `pfo->lpVtbl->DeleteItem(pfo, pSI, NULL);` to increment the reference count. However, even when this is plausible and sensible, I would rather not assume behavior that is not clearly document.

Based on

**Comment by robertguetzkow** [comment]
Yes, that is obviously the idea behind the reference counting for COM. Microsoft provides the following two articles:

- https://docs.microsoft.com/en-us/windows/win32/com/rules-for-managing-reference-counts
- https://docs.microsoft.com/en-us/windows/win32/learnwin32/managing-the-lifetime-of-an-object

Based on them, it is indeed clearly documented that the functions being called are responsible for handling the reference count based on what they do with the pointer. However, I have also checked the return value of `pSI->lpVtbl->Release(pSI);` when it is called directly after `pfo->lpVtbl->DeleteItem(pfo, pSI, NULL);` and the reference count is zero. Apparently, there is no reference kept to the original for the purpose of conducting the operation later on. Instead it probably allocates memory and creates a copy of the relevant information. 

In summary, it seems that you are correct that `pSI->lpVtbl->Release(pSI);` can be safely called prior to `pfo->lpVtbl->PerformOperations(pfo);`.

**Comment by robertguetzkow** [comment]
In order to keep the error handling and control flow consistent, I will keep the extra label. The implementation and design of libtrashcan are far from perfect and the current code formatting is strange, but I hope that this good enough for an obscure library that (almost) no one uses. I may have to revisit this project from the ground up if people are genuinely interested in it.

Thank you very much for your input and report! This both eliminates a memory leak and led to the identification of another bug (#12). More importantly, these issues also apply to Blender where I will patch them as well.
